### PR TITLE
Sizing speed up related to reducing the number of warmup days for design days

### PR DESCRIPTION
**Pull request by EnergyArchmage**
Change minimum warmup days default from 6 to 1. In the Building input object, the input field called Minimum Number of Warmup Days had a default value of 6 days.  This changes the IDD default value from 6 to 1 without harm.  If the warmup converges fast based on convergence criteria then great, quit.  If the minimum is not enough, fine, it will still go up to the maximum number of days.  Combined with the thermal history reset suppress option (below) we do see very fast convergence on warm up days.

Add feature to allow suppressing thermal reset at beginning of design days.  Similar environmental periods can reuse the thermal history from the previous day to speed up warm up. For example a series of cooling design days can run consecutively without re initializing to 23C and warmup much faster. 

Add feature to fine control maximum number of warmup days for design day environments. Users may want to tolerate much lower maximum number of days for design days with minimal impact on sizing outcomes.

These changes are contributed by TRANE in discussion with 10X project. 

Pull request overview
---------------------
 - Fixes #ISSUENUMBERHERE (IF THIS IS A DEFECT)
 - DESCRIBE PURPOSE OF THIS PULL REQUEST

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@EnergyArchmage @lgentile it has been 28 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
This actually looks really good.  I pulled in develop and changed a unit test flow rate by about 1e-6 to get the unit test passing.  Other than the unit test issue the only diffs were a handful of EIO and table diffs that were purely due to the change in warmup days.  I'm running regressions locally since CI is bogged down, and I still need to do the code review, but I am hopeful.

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -482,11 +482,12 @@ Building,
        \key Urban
        \default Suburbs
   N2 , \field Loads Convergence Tolerance Value
-       \note Loads Convergence Tolerance Value is a fraction of load
+       \note Loads Convergence Tolerance Value is a change in load from one warmup day to the next
        \type real
        \minimum> 0.0
        \maximum .5
        \default .04
+       \units W
```
Comment (shorowit) [inline idd/Energy+.idd.in:490]: So the code was already interpreting this input as W, but it was described as a fraction? If so, it appears that the E+ documentation needs to be updated to reflect this.

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -1155,13 +1158,26 @@ SizingPeriod:DesignDay,
...
        \note Used if Sky Model Indicator = ASHRAEClearSky or ZhangHuang
        \minimum 0.0
        \maximum 1.2
        \default 0.0
        \note 0.0 is totally unclear, 1.0 is totally clear
        \type real
+  N15, \field Maximum Number Warmup Days 
+       \note If used this design day will be run with a custom limit on the maximum number of days that are repeated for warmup.
+       \note Limiting the number of warmup days can improve run time.
+       \note TRACE(R) 3D Plus custom feature.
```
Comment (mjwitte) [inline idd/Energy+.idd.in:1171]: @Myoldmopar I think we want to delete this note? And same for 1176.

**Comment by Myoldmopar** [comment]
Local diff summary:
--------------
1ZoneUncontrolledResLayers 
 - EIO diff that was purely the number of warmup days
 - Table diff that was purely the number of warmup days

RefrigeratedWarehouse
 - EIO diff that was purely the number of warmup days
 - Table diff that was purely the number of warmup days

ConstSpeedBranchPumpExercise
 - EIO diff that was purely the number of warmup days

AirflowNetworkAdvanced_SingleSided_NV
 - EIO diffs including the number of warmup days
 - EIO diffs include warmup convergence, but it's really no difference:

image

image
 - ESO diffs that appear to be significant.  Most of the values are pretty small, but for some reason, the zone temps are consistently off on the September day by 0.2C, like all day long.  That's too far to be a small diff unless this fixed something.  

The rest of the files are perfect, no diffs, so I'm wondering if there is some corner case with this airflow network file that needs a special case somewhere in the code changes.

File: idd/Energy+.idd.in (hunk 2 above)
Comment (Myoldmopar) [reply idd/Energy+.idd.in:1171]: Right, this should go away

File: idd/Energy+.idd.in (hunk 1 above)
Comment (Myoldmopar) [reply idd/Energy+.idd.in:490]: Yeah if this was described as a fraction in the docs then they need to be changed.

File: idd/Energy+.idd.in [hunk 3]
```diff
@@ -1155,13 +1158,26 @@ SizingPeriod:DesignDay,
...
+       \type integer
+  A13; \field Begin Environment Reset Mode
+       \note If used this can control if you want the thermal history to be reset at the beginning of the design day.
+       \note When using a series of similiar design days, this field can be used to retain warmup state from the previous design day.
+       \note TRACE(R) 3D Plus custom feature.
```
Comment (EnergyArchmage) [inline idd/Energy+.idd.in:1176]: missed it here

File: src/EnergyPlus/SimulationManager.cc [hunk 4]
```diff
@@ -497,6 +497,14 @@ namespace SimulationManager {
             DisplayString("Initializing New Environment Parameters");
 
             BeginEnvrnFlag = true;
+            if ((KindOfSim == ksDesignDay) && (WeatherManager::DesDayInput(Environment(Envrn).DesignDayNum).suppressBegEnvReset)) {
+                // user has input in SizingPeriod:DesignDay directing to skip begin environment rests, for accuracy-with-speed as zones can more
+                // easily converge fewer warmup days are allowed
+                DisplayString("TRACE(R) 3D Plus Mode: Suppressing Initialization of New Environment Parameters");
```
Comment (EnergyArchmage) [inline src/EnergyPlus/SimulationManager.cc]: need to clean here as well

**Comment by Myoldmopar** [comment]
@EnergyArchmage will you have a chance to look at the one file with large ESO diffs?

File: src/ConvertInputFormat/main.cpp [hunk 5]
```diff
@@ -497,11 +496,10 @@ int main(int argc, const char *argv[])
 #endif
 
 #ifdef _OPENMP
-#pragma omp parallel default(none)                                                                                                                   \
-    shared(files, number_files, fileCount, schema, outputType, outputTypeStr, output_directory)
+#pragma omp parallel default(none) shared(files, number_files, fileCount, schema, outputType, outputTypeStr, output_directory)
     {
 #pragma omp for
-        for (std::size_t i = 0; i < number_files; ++i) {
+        for (int i = 0; i < number_files; ++i) {
```
Comment (EnergyArchmage) [inline src/ConvertInputFormat/main.cpp]: oops, didn't mean to commit this.  I had to do this to get it to compile in VS2017

**Comment by Myoldmopar** [comment]
@EnergyArchmage the custom_check results was because the branch was out of date (license year).  Also, I'm not sure whether Windows failures were valid or some sort of fluke.  I'll pull develop into this branch and let CI rerun this afternoon and we'll see where we are actually at.

**Comment by Myoldmopar** [comment]
OK, CI looks to be happy again.  Just the one airflow network file with large ESO diffs.  It looks like those diffs have been there since the very beginning of this PR.  Do we need to modify the IDF to force that file back into the correct number of warmup days?

**Comment by EnergyArchmage** [comment]
The test files that show diffs leave the input field blank. So now they get the new default of 1 instead of 6.  Then they only need 2 warmup days instead of 6.  I can only guess that the airflow network model is just in a slightly different state at the beginning of the simulation because of the shorter warmup. 

The idf files showing diffs could be changed to set min days at 6 if desired.   On the other hand the vast majority of files that have input set at 6 are not affected by the changes here and so the performance speed up is suppressed.

**Comment by Myoldmopar** [comment]
OK, I set the file to a minimum of 6 warmup days and got no diffs.  I pushed that change up along with an update from the develop branch.  I believe this branch will be able to drop as soon as CI is complete this afternoon.  Thanks @EnergyArchmage !

**Comment by Myoldmopar** [comment]
CI looks _very_ happy now.  If it continues that way, this can go in.  Thanks @EnergyArchmage

**Comment by Myoldmopar** [comment]
OK, calling it.  This is good to go.  Thanks @EnergyArchmage

**Pull request merged**